### PR TITLE
Object whats here

### PR DESCRIPTION
**Pull request by BurdetteLamar**
I know that clone and dup are different, but I don't recall how.

Also, the links to other What's Here sections cite the .html files.  I think this is suboptimal, but could not get them to work otherwise.

### CODE REVIEW

**Comment by BurdetteLamar** [comment]
@jeremyevans, FYI, I have Numeric and Comparable ready for the next PR, but will not be writing from Wednesday through Saturday.

File: object.c [hunk 1]
```diff
@@ -4532,6 +4532,95 @@ f_sprintf(int c, const VALUE *v, VALUE _)
...
+ *                                    +self+ is an instance of a superclass
+ *                                    of class +name+,
+ *                                    or +self+ includes module +name+.
+ *  - #remove_instance_variable:: Removes the named instance variable from +self+.
+ *
+ *  === Other
+ *
+ *  - ::const_missing:: Called when an uninitialized constant is referenced.
+ *  - ::receive_if:: Receive only a specific message.
+ *  - #clone::  Returns a shallow copy of +self+.
```
Comment (jeremyevans) [inline object.c:4611]: ... `including singleton class and frozen state.`

File: object.c [hunk 2]
```diff
@@ -4532,6 +4532,95 @@ f_sprintf(int c, const VALUE *v, VALUE _)
...
+ *  - #display:: Prints +self+ to the given \IO stream.
+ *  - #dup:: Returns a shallow copy of +self+.
```
Comment (jeremyevans) [inline object.c:4613]: `shallow` -> `shallow unfrozen`

File: object.c [hunk 3]
```diff
@@ -4532,6 +4532,95 @@ f_sprintf(int c, const VALUE *v, VALUE _)
...
+ *                                    of class +name+,
+ *                                    or +self+ includes module +name+.
+ *  - #remove_instance_variable:: Removes the named instance variable from +self+.
+ *
+ *  === Other
+ *
+ *  - ::const_missing:: Called when an uninitialized constant is referenced.
+ *  - ::receive_if:: Receive only a specific message.
+ *  - #clone::  Returns a shallow copy of +self+.
+ *  - #display:: Prints +self+ to the given \IO stream.
```
Comment (jeremyevans) [inline object.c:4612]: ... `or $stdout`.

File: object.c [hunk 4]
```diff
@@ -4532,6 +4532,95 @@ f_sprintf(int c, const VALUE *v, VALUE _)
...
+ *  - #dup:: Returns a shallow copy of +self+.
+ *  - #enum_for:: Returns an Enumerator for +self+ using the given method,
```
Comment (jeremyevans) [inline object.c:4614]: ... `using the given method, arguments, and block`

File: object.c [hunk 5]
```diff
@@ -4532,6 +4532,95 @@ f_sprintf(int c, const VALUE *v, VALUE _)
...
+ *                optional arguments, and optional block.
+ *  - #freeze:: Prevents further modifications to +self+.
+ *  - #hash:: Returns the integer hash value for +self+.
+ *  - #inspect:: Returns a string representation of +self+.
+ *  - #itself:: Returns +self+.
+ *  - #to_enum:: Returns an Enumerator for +self+ using the given method,
```
Comment (jeremyevans) [inline object.c:4620]: `to_enum` is an alias of `enum_for`.

File: object.c [hunk 6]
```diff
@@ -4532,6 +4532,95 @@ f_sprintf(int c, const VALUE *v, VALUE _)
...
+ *  - ::const_missing:: Called when an uninitialized constant is referenced.
+ *  - ::receive_if:: Receive only a specific message.
+ *  - #clone::  Returns a shallow copy of +self+.
+ *  - #display:: Prints +self+ to the given \IO stream.
+ *  - #dup:: Returns a shallow copy of +self+.
+ *  - #enum_for:: Returns an Enumerator for +self+ using the given method,
+ *                optional arguments, and optional block.
+ *  - #freeze:: Prevents further modifications to +self+.
+ *  - #hash:: Returns the integer hash value for +self+.
+ *  - #inspect:: Returns a string representation of +self+.
```
Comment (jeremyevans) [inline object.c:4618]: `string` -> `human readable string` (to distinguish from `to_s`)

**Review by jeremyevans — CHANGES_REQUESTED** [review]
Thanks for working on this! Please see inline comments.

File: object.c [hunk 7]
```diff
@@ -4532,6 +4532,95 @@ f_sprintf(int c, const VALUE *v, VALUE _)
...
+ *  - #kind_of? (aliased as #is_a?):: Returns +true+ if, for the given +name+,
+ *                                    +self+ is an instance of class +name+,
+ *                                    +self+ is an instance of a superclass
+ *                                    of class +name+,
+ *                                    or +self+ includes module +name+.
+ *  - #remove_instance_variable:: Removes the named instance variable from +self+.
+ *
+ *  === Other
+ *
+ *  - ::const_missing:: Called when an uninitialized constant is referenced.
```
[the remark this answers is not in the document]
Comment (BurdetteLamar) [reply object.c:4609]: So remove?

File: object.c [hunk 8]
```diff
@@ -4532,6 +4532,95 @@ f_sprintf(int c, const VALUE *v, VALUE _)
...
+ *  - ::receive_if:: Receive only a specific message.
```
[the remark this answers is not in the document]
Comment (BurdetteLamar) [reply object.c:4610]: So remove?

File: object.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (jeremyevans) [reply object.c:4602]: Yes, same for `receive_if`.

**Comment by chrisseaton** [comment]
Aren't these methods already documented? For example what `instance_eval` is now described in two places - `vm_eval.c` and `object.c`. Are we going to make it harder to keep this documentation in sync?

**Comment by BurdetteLamar** [comment]
> 
> 
> Aren't these methods already documented? For example what `instance_eval` is now described in two places - `vm_eval.c` and `object.c`. Are we going to make it harder to keep this documentation in sync?

#instance_eval and #instance_exec are not in Object doc 3.0.1, so I don't know how they got on my list for the What's Here.  Should these two be removed?  Are there others?

**Comment by chrisseaton** [comment]
I think I'm more fundamentally confused what you're documenting here - `Object` doesn't define any methods at all!

```
irb(main):001:0> Object.methods(false)
=> []
```

And more generally we're now describing what each method does in two places - and in this case in confusing places that don't actually define the method. Now two places to change if the method changes, and you may not know to look here to do it. Can't we build these summary blocks of what method is in each class automatically? The tools already give you a list of methods - could we annotate the 'summary phrase' short version that you're adding as metadata in the main method documentation and then collate them using a tool?

**Comment by jeremyevans** [comment]
@chrisseaton Please see the top of https://docs.ruby-lang.org/en/master/Object.html.  It describes that the methods documented for Object are actually implemented in Kernel.  I'm not sure the historical reason for this, but it is deliberately this way.  I would be in favor of documenting everything in Kernel.

In terms of documenting in the same place as the method definition, currently rdoc doesn't support that.  If supported is added to rdoc to enable that, we could potentially use it, but as it doesn't currently exist, that is not currently an option. The only alternative currently would be not having a "What's Here" at all.

**Comment by BurdetteLamar** [comment]
> 
> 
> I think I'm more fundamentally confused what you're documenting here - `Object` doesn't define any methods at all!
> 
> ```
> irb(main):001:0> Object.methods(false)
> => []
> ```
> 
> And more generally we're now describing what each method does in two places - and in this case in confusing places that don't actually define the method. Now two places to change if the method changes, and you may not know to look here to do it. Can't we build these summary blocks of what method is in each class automatically? The tools already give you a list of methods - could we annotate the 'summary phrase' short version that you're adding as metadata in the main method documentation and then collate them using a tool?

@chrisseaton  For all technical matters, I'm relying on technical reviews by @jeremyevans and others.  As for your last point, I think that would be *much* better, but it's far out-of-scope for me.

**Comment by jeremyevans** [comment]
@BurdetteLamar `instance_exec` and `instance_eval` you documented in "What's Here" for `BasicObject`, which is correct as that is where they are defined.

**Comment by BurdetteLamar** [comment]
> 
> 
> @BurdetteLamar `instance_exec` and `instance_eval` you documented in "What's Here" for `BasicObject`, which is correct as that is where they are defined.
Now removed from Object's What's Here.

**Comment by jeremyevans** [comment]
> > @BurdetteLamar `instance_exec` and `instance_eval` you documented in "What's Here" for `BasicObject`, which is correct as that is where they are defined.
> 
> @jeremyevans, now removed from Object's What's Here.

No, you removed the methods from BasicObject's What's Here.  As I said, those methods were correctly documented in BasicObject, so there was no need to remove them.

File: object.c [hunk 9]
```diff
@@ -4532,6 +4529,90 @@ f_sprintf(int c, const VALUE *v, VALUE _)
...
+ *
+ *  First, what's elsewhere. \Class \Object:
+ *
+ *  - Inherits from {class BasicObject}[BasicObject.html#class-BasicObject-label-What-27s+Here].
+ *  - Includes {module Kernel}[Kernel.html#module-Kernel-label-What-27s+Here].
+ *
+ *  Here, class \Object provides methods for:
+ *
+ *  - {Querying}[#class-Object-label-Querying]
+ *  - {Methods}[#class-Object-label-Methods]
```
Comment (jeremyevans) [inline object.c:4543]: The `Methods` section was removed.

**Review by jeremyevans — CHANGES_REQUESTED** [review]
Looks good. One small new comment, and the previous comment for #clone still needs to be addressed, then it should be ready.

**Review by jeremyevans — APPROVED** [review]
Looks good!

**Comment by BurdetteLamar** [comment]
@jeremyevans, I am getting emails about failed checks against this PR.  I can't make out what's wrong.  Some checks were cancelled.  We were clean when I merged 4 hours ago.

**Comment by jeremyevans** [comment]
You can ignore CI failures for documentation only commits.

**Pull request merged**